### PR TITLE
feat(fn-416): cosmosdb

### PR DESCRIPTION
**Pull request by OxonianCambion**
## Introduction
We need to be able to manage the infrastructure in declarative IaC - we've chosen Bicep.
The first phase is to be able to:

- produce a complete new deployment environment (feature) mirroring the current ones.
- be in a position to take over the extant environments.

Note that there have been some manual changes to the environments. This work also functions as an audit of the current infrastructure.
Possible enhancements / updates will be noted but not implemented in the first phase.

This card covers wiring up cosmosdb (MongoDB) only. Note that there will be some tidying / refactoring when all the components are in place, (e.g. commonise tags, consider parameter sets), but the code should be of reasonable quality and clear.
### Resolution

This code wires up cosmosdb and its private link.

### CODE REVIEW

File: infrastructure/arm_templates/cosmosdb/errors.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+    "code": "ExportTemplateCompletedWithErrors",
```
Comment (OxonianCambion) [inline infrastructure/arm_templates/cosmosdb/errors.json:2]: Skip over the json files.

This one is a bit different in that it is the warnings about what couldn't be exported. However, none of these resources are in the CLI scripts, so I think we're ok!

File: infrastructure/resource_group_level/modules/cosmosdb.bicep [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+    ]
+    cors: []
+    capabilities: capabilities
+    ipRules: [for ip in allowedIps: {
+      ipAddressOrRange: ip
+    }]
+    backupPolicy: {
+      type: 'Continuous'
+      continuousModeProperties: {
+        tier: 'Continuous30Days'
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/cosmosdb.bicep:114]: Note that this is paid for. I think we should probably use the free 7 days backup for all environments apart from live.
Agreed? If so, I'll add it in.

**Comment by abhi-markan** [comment]
> _ No description provided. _

Missing description can we use following format please:

```
## Introduction
## Resolution
```

File: infrastructure/resource_group_level/main.bicep [hunk 3]
```diff
@@ -57,6 +57,9 @@ param onPremiseNetworkIps array = [
 param storageNetworkAccessDefaultAction string= 'Allow'
 // Staging has the default as Deny, corresponding to "Enabled from selected virtual networks and IP addresses".
 
+@allowed(['Provisioned Throughput', 'Serverless'])
+param cosmosDbCapacityMode string = 'Serverless' // All current environments use 'Provisioned Throughput'
```
[the remark this answers is not in the document]
Comment (IgnatG) [reply infrastructure/resource_group_level/main.bicep:61]: serverless will involve a cold start. the codebase (nodeJS part) is not ready for it. let's leave it as it is for now - Provision Throughput. we can modify it later on if needed

File: infrastructure/resource_group_level/modules/cosmosdb.bicep [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+@allowed(['Provisioned Throughput', 'Serverless'])
+param capacityMode string
+
+var cosmosDbName = 'tfs-${environment}-mongo'
+var privateEndpointName = 'tfs-${environment}-mongo'
+
+
+// TODO:FN-416 Work out what these misc. IPs with access are.
+// Misc IPs
+var miscIps =  [
```
[the remark this answers is not in the document]
Comment (IgnatG) [reply infrastructure/resource_group_level/modules/cosmosdb.bicep:18]: all those IPs can be removed

File: infrastructure/resource_group_level/modules/cosmosdb.bicep (hunk 2 above)
Comment (IgnatG) [reply infrastructure/resource_group_level/modules/cosmosdb.bicep:114]: yup, good shout

**Review by IgnatG — APPROVED** [review]
All good overall. we can change it to serverless later though - let's focus on like-for-like migration first

File: infrastructure/resource_group_level/main.bicep (hunk 3 above)
[the remark this answers is not in the document]
Comment (OxonianCambion) [reply infrastructure/resource_group_level/main.bicep:61]: This is (badly named) serverless mode for CosmosDB. I prefer to thing it as consumption mode.
I don't think there's any cold start issues for CosmosDB.

At a minimum, I'd probably use serverless for dev / test, if it wasn't for the fact that they already exist. (And you can't just conver it.) To be fair, the cost for the current capacity is $1 a day but serverless will be essentially free.

In summary, I'll use `Serverless` for my `feature` environment and `Provisioned Throughput` for all the others.

[the remark this answers is not in the document]
Comment (IgnatG) [reply infrastructure/resource_group_level/main.bicep]: Ok cool

**Pull request merged**